File: frontend/lib/types/material.ts
/**
 * Domain types for SIH 2026 Unified Material Intelligence Platform (Phase 1).
 * Domain types for SIH 2026 Unified Material Intelligence Platform.
 */

export interface IngestionErrorItem {
  row: number;
  legacy_code?: string | null;
  reason: string;
}

export interface MaterialIngestionResponse {
  ingestion_id: string;
  filename: string;
  total_records: number;
  accepted_records: number;
  rejected_records: number;
  duplicate_records: number;
  errors: IngestionErrorItem[];
}

export interface MaterialRecord {
  id: string;
  organization: string;
  legacy_code: string;
  description: string;
  uom?: string | null;
  item_name?: string | null;
  part_number?: string | null;
  manufacturer?: string | null;
  equipment_compatibility?: string | null;
  material_type?: string | null;
  category?: string | null;
  specification?: string | null;
  national_id?: string | null;
  status?: string | null;
  source_file?: string | null;
  source_row?: number | null;
  probability_score?: number | null;
  created_at: string;
  updated_at: string;
}

export interface PaginatedMaterialsResponse {
  items: MaterialRecord[];
  total: number;
  limit: number;
  offset: number;
  total_pages?: number;
  page?: number;
  page_size?: number;
}

export interface MaterialAttribute {
  name: string;
  value: string | number | boolean;
  unit?: string | null;
  confidence?: number | null;
}

export interface MaterialItem {
  id: string;
  itemCode: string;
  description: string;
  organizationId?: string | null;
  organizationName?: string | null;
  category?: string | null;
  attributes: MaterialAttribute[];
  createdAt?: string;
  updatedAt?: string;
}

export interface EquivalentMatch {
  id: string;
  sourceItemId: string;
  targetItemId: string;
  sourceItem?: MaterialItem;
  targetItem?: MaterialItem;
  similarityScore: number;
  matchStatus: "pending" | "verified" | "rejected";
  matchedAttributes: string[];
  rationale?: string | null;
  confidence: number;
}

export interface VerificationReview {
  id: string;
  matchId: string;
  reviewerId: string;
  reviewerName?: string;
  decision: "approved" | "rejected" | "modified";
  comments?: string | null;
  timestamp: string;
}

export interface CPSEOrganization {
  id: string;
  name: string;
  code: string;
  sector?: string;
  catalogCount?: number;
}

export interface MaterialQualityStats {
  total_materials: number;
  organizations: number;
  missing_uom: number;
  missing_manufacturer: number;
  missing_part_number: number;
  missing_category: number;
  assigned_count?: number;
  unassigned_count?: number;
}

File: backend/api-service/app/schemas/material.py
"""Foundational domain schemas for Unified Material Intelligence."""

import uuid
from datetime import datetime
from typing import Any
from pydantic import BaseModel, ConfigDict, Field



class MaterialAttribute(BaseModel):
    """Normalized technical attribute key-value pair."""

    name: str
    value: str | float | int | bool
    unit: str | None = None
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class IngestionErrorItem(BaseModel):
    """Details for a single record validation failure or duplicate error."""

    row: int
    legacy_code: str | None = None
    reason: str


class MaterialIngestionResponse(BaseModel):
    """Result summary of a batch material ingestion operation."""

    ingestion_id: str
    filename: str
    total_records: int
    accepted_records: int
    rejected_records: int
    duplicate_records: int
    errors: list[IngestionErrorItem] = Field(default_factory=list)


class MaterialRecordInput(BaseModel):
    """Standardized internal representation of an ingested material record."""

    organization: str
    legacy_code: str
    description: str
    uom: str | None = None
    item_name: str | None = None
    part_number: str | None = None
    manufacturer: str | None = None
    equipment_compatibility: str | None = None
    material_type: str | None = None
    category: str | None = None
    specification: str | None = None
    national_id: str | None = None
    status: str | None = "AVAILABLE"
    source_file: str | None = None
    source_row: int | None = None


class MaterialResponse(BaseModel):
    """API response model for a single persisted material record."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization: str
    legacy_code: str
    description: str
    uom: str | None = None
    item_name: str | None = None
    part_number: str | None = None
    manufacturer: str | None = None
    equipment_compatibility: str | None = None
    material_type: str | None = None
    category: str | None = None
    specification: str | None = None
    national_id: str | None = None
    status: str | None = "AVAILABLE"
    source_file: str | None = None
    source_row: int | None = None
    created_at: datetime
    updated_at: datetime


class MaterialListResponse(BaseModel):
    """Paginated list of material catalog items."""

    items: list[MaterialResponse]
    total: int
    limit: int
    offset: int
    page: int = 1
    page_size: int = 20
    total_pages: int = 1


class MaterialQualityStats(BaseModel):
    """Deterministic calculation of completeness metrics across all material master records."""

    total_materials: int
    organizations: int
    missing_uom: int
    missing_manufacturer: int
    missing_part_number: int
    missing_category: int
    assigned_count: int = 0
    unassigned_count: int = 0



class MaterialBase(BaseModel):
    """Core material entity foundation."""

    id: str | None = None
    item_code: str
    description: str
    organization_id: str | None = None
    category: str | None = None
    attributes: list[MaterialAttribute] = Field(default_factory=list)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class EquivalentMatchStub(BaseModel):
    """Match recommendation between material items."""

    source_item_id: str
    target_item_id: str
    similarity_score: float = Field(ge=0.0, le=1.0)
    match_status: str = "pending"  # "pending", "verified", "rejected"
    matched_attributes: list[str] = Field(default_factory=list)
    rationale: str | None = None


class ReviewStub(BaseModel):
    """Verification review record for material harmonization."""

    id: str | None = None
    match_id: str
    reviewer_id: str
    decision: str  # "approved", "rejected", "modified"
    comments: str | None = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)

